In [1]:
import pandas as pd

# Load the data
data = pd.read_csv('../data/MachineLearningRating_v3.txt', delimiter='|', low_memory=False) 


In [2]:
data.head()

UnderwrittenCoverID  PolicyID     TransactionMonth  IsVATRegistered  \
0               145249     12827  2015-03-01 00:00:00             True   
1               145249     12827  2015-05-01 00:00:00             True   
2               145249     12827  2015-07-01 00:00:00             True   
3               145255     12827  2015-05-01 00:00:00             True   
4               145255     12827  2015-07-01 00:00:00             True   

  Citizenship          LegalType Title Language                 Bank  \
0              Close Corporation    Mr  English  First National Bank   
1              Close Corporation    Mr  English  First National Bank   
2              Close Corporation    Mr  English  First National Bank   
3              Close Corporation    Mr  English  First National Bank   
4              Close Corporation    Mr  English  First National Bank   

       AccountType  ...                    ExcessSelected CoverCategory  \
0  Current account  ...             Mobility - Windscreen    Windscreen   
1  Current account  ...             Mobility - Windscreen    Windscreen   
2  Current account  ...             Mobility - Windscreen    Windscreen   
3  Current account  ...  Mobility - Metered Taxis - R2000    Own damage   
4  Current account  ...  Mobility - Metered Taxis - R2000    Own damage   

    CoverType            CoverGroup              Section  \
0  Windscreen  Comprehensive - Taxi  Motor Comprehensive   
1  Windscreen  Comprehensive - Taxi  Motor Comprehensive   
2  Windscreen  Comprehensive - Taxi  Motor Comprehensive   
3  Own Damage  Comprehensive - Taxi  Motor Comprehensive   
4  Own Damage  Comprehensive - Taxi  Motor Comprehensive   

                           Product StatutoryClass StatutoryRiskType  \
0  Mobility Metered Taxis: Monthly     Commercial     IFRS Constant   
1  Mobility Metered Taxis: Monthly     Commercial     IFRS Constant   
2  Mobility Metered Taxis: Monthly     Commercial     IFRS Constant   
3  Mobility Metered Taxis: Monthly     Commercial     IFRS Constant   
4  Mobility Metered Taxis: Monthly     Commercial     IFRS Constant   

   TotalPremium TotalClaims  
0     21.929825         0.0  
1     21.929825         0.0  
2      0.000000         0.0  
3    512.848070         0.0  
4      0.000000         0.0  

[5 rows x 52 columns]

In [3]:
# Get a concise summary of the dataset
print(data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000098 entries, 0 to 1000097
Data columns (total 52 columns):
 #   Column                    Non-Null Count    Dtype  
---  ------                    --------------    -----  
 0   UnderwrittenCoverID       1000098 non-null  int64  
 1   PolicyID                  1000098 non-null  int64  
 2   TransactionMonth          1000098 non-null  object 
 3   IsVATRegistered           1000098 non-null  bool   
 4   Citizenship               1000098 non-null  object 
 5   LegalType                 1000098 non-null  object 
 6   Title                     1000098 non-null  object 
 7   Language                  1000098 non-null  object 
 8   Bank                      854137 non-null   object 
 9   AccountType               959866 non-null   object 
 10  MaritalStatus             991839 non-null   object 
 11  Gender                    990562 non-null   object 
 12  Country                   1000098 non-null  object 
 13  Province                  1

In [4]:
numerical_columns = data.select_dtypes(include=['number']).columns
print("Descriptive Statistics for Numerical Features:")
print(data[numerical_columns].describe())

Descriptive Statistics for Numerical Features:
       UnderwrittenCoverID      PolicyID    PostalCode        mmcode  \
count         1.000098e+06  1.000098e+06  1.000098e+06  9.995460e+05   
mean          1.048175e+05  7.956682e+03  3.020601e+03  5.487770e+07   
std           6.329371e+04  5.290039e+03  2.649854e+03  1.360381e+07   
min           1.000000e+00  1.400000e+01  1.000000e+00  4.041200e+06   
25%           5.514300e+04  4.500000e+03  8.270000e+02  6.005692e+07   
50%           9.408300e+04  7.071000e+03  2.000000e+03  6.005842e+07   
75%           1.391900e+05  1.107700e+04  4.180000e+03  6.005842e+07   
max           3.011750e+05  2.324600e+04  9.870000e+03  6.506535e+07   

       RegistrationYear      Cylinders  cubiccapacity      kilowatts  \
count      1.000098e+06  999546.000000  999546.000000  999546.000000   
mean       2.010225e+03       4.046642    2466.743258      97.207919   
std        3.261391e+00       0.294020     442.800640      19.393256   
min        1.987

In [2]:
import sys
import os
sys.path.append(os.path.abspath('../scripts'))

# Get the current directory
current_dir = os.getcwd()
print(current_dir)

# Get the parent directory
parent_dir = os.path.dirname(current_dir)
print(parent_dir)

# Insert the path to the parent directory
sys.path.insert(0, parent_dir)
from scripts.descriptive_statstics import Stat


c:\Users\HP\Documents\10-Academy\AlphaCare-Insurance-Solutions-WEEK3\notebooks
c:\Users\HP\Documents\10-Academy\AlphaCare-Insurance-Solutions-WEEK3


In [3]:
analyzer = Stat(data)

In [4]:
# to print the decriptive analysis for total perimum column
print("Descriptive Statistics for total perimum:")
analyzer.decriptive_stat(['TotalPremium'])

Descriptive Statistics for total perimum:


{'TotalPremium': {'Mean': 61.90549595782839,
  'Standard Deviation': 230.28451275733377,
  'Variance': 53030.956815882615,
  'Range': 66065.18017543857,
  'Interquartile Range (IQR)': 21.929824561403}}

In [14]:
# to print the decriptive analysis for total claim column
print("Descriptive Statistics for total claims:")
analyzer.decriptive_stat(['TotalClaims'])

Descriptive Statistics for total claims:


{'TotalClaims': {'Mean': 64.86118977410388,
  'Standard Deviation': 2384.0746945027145,
  'Variance': 5683812.14896821,
  'Range': 405094.5175438598,
  'Interquartile Range (IQR)': 0.0}}

Wide Spread in Both Premiums and Claims: Both TotalPremium and TotalClaims show high variability, with large ranges and standard deviations. This suggests that your dataset contains both very low and very high values for premiums and claims, likely due to a few outliers.

Extreme Claims: The claims have much higher variability and range compared to the premiums. This could indicate that while premiums are somewhat consistent, the claims data has a few extreme values that are much higher than the average.

Potential Outliers: The large standard deviations and ranges in both columns suggest the presence of outliers.

In [4]:
analyzer.review_data_structure()

Data Types of Each Column:
UnderwrittenCoverID           int64
PolicyID                      int64
TransactionMonth             object
IsVATRegistered                bool
Citizenship                  object
LegalType                    object
Title                        object
Language                     object
Bank                         object
AccountType                  object
MaritalStatus                object
Gender                       object
Country                      object
Province                     object
PostalCode                    int64
MainCrestaZone               object
SubCrestaZone                object
ItemType                     object
mmcode                      float64
VehicleType                  object
RegistrationYear              int64
make                         object
Model                        object
Cylinders                   float64
cubiccapacity               float64
kilowatts                   float64
bodytype                     object
N

Data Types Review:

The dataset consists of a mix of numerical (e.g., int64, float64) and categorical/object data types (e.g., TransactionMonth, LegalType).
Numeric Columns: Columns like UnderwrittenCoverID, PolicyID, PostalCode, and SumInsured contain numeric values.
Categorical/Object Columns: Many of the columns such as Citizenship, Bank, MaritalStatus, and CoverType are treated as object types, which likely represent categorical variables.
Potential Categorical Variables:

There are 35 columns that are potential categorical variables, indicating that a large portion of the dataset contains non-numeric, categorical information. These include personal characteristics (e.g., Citizenship, Gender), insurance-related details (e.g., CoverType, CoverGroup), and financial details (e.g., Bank, AccountType).
These categorical columns will need to be handled carefully during any modeling or analytical process, likely through encoding techniques like one-hot encoding or label encoding.
Potential DateTime Columns:

No columns were automatically identified as likely DateTime fields, even though some columns like TransactionMonth or VehicleIntroDate might contain date-like information. This may indicate inconsistencies in the data format, or these fields may not be in a format that can be easily parsed by the date detection algorithm.

In [4]:
analyzer.check_missing_values()


Missing Values Summary:
NumberOfVehiclesInFleet    1000098
CrossBorder                 999400
CustomValueEstimate         779642
Converted                   641901
Rebuilt                     641901
WrittenOff                  641901
NewVehicle                  153295
Bank                        145961
AccountType                  40232
Gender                        9536
MaritalStatus                 8259
Cylinders                      552
cubiccapacity                  552
kilowatts                      552
NumberOfDoors                  552
VehicleIntroDate               552
Model                          552
make                           552
VehicleType                    552
mmcode                         552
bodytype                       552
CapitalOutstanding               2
dtype: int64


The dataset has several columns with missing values. Notably, NumberOfVehiclesInFleet is entirely missing, and CrossBorder is missing in 999,400 rows. Other columns with significant missing values include CustomValueEstimate (779,642 rows) and Converted, Rebuilt, and WrittenOff (641,901 rows each). Additionally, fields like Bank, AccountType, Gender, and MaritalStatus also have missing data, although to a lesser extent. Key vehicle-related fields like Cylinders, cubiccapacity, and make have 552 missing entries.